agent/src/com/cloud/agent/Agent.java

```diff
@@ -990,4 +1091,53 @@ public void doTask(final Task task) throws TaskExecutionException {
             }
         }
     }
+
+    public class HostLBCheckerTask extends ManagedContextTimerTask {
+
+         @Override
+         protected void runInContext() {
+             try {
+                 int index = 0;
+                 final String[] msList = _shell.getHosts();
+                 String preferredHost  = msList[index];
+                 final String connectedHost = _shell.getConnectedHost();
+                 if (s_logger.isTraceEnabled()) {
+                     s_logger.trace("Running host LB checker task, connected host=" + connectedHost + ", preferred host=" + preferredHost);
+                 }
+                 final long wait = _shell.getFailbackPollingWait();
+                 while (!preferredHost.equals(connectedHost)) {
```

Comment:
Let's not have a while loop here. @nvazquez if you're not aware, when the agent is disconnected the `reconnect()` logic already kicks and is responsible for performing reconnections. This timer/task should not share the same responsibility, instead it should check if the preferred host (first in the list) is available and if the agent is already connected, only then it should disconnect by performing a reconnect on the `_link` (if it's not null for example).

Bear in mind, we've two assumptions here (1) the first host in the list is the preferred one, (2) the timer task is responsible for reconnecting to the preferred host.

Reply:
Done, thanks